### ISSUE TITLE
Submission: Group 11: Car Evaluation Analysis

### ISSUE THREAD

**Issue by davyxuximin** [UBC-MDS/data-analysis-review-2024#15]
**Submitting authors:**  @nvarabioff @mdkbmi @zze1999 @davyxuximin

**Repository: https://github.com/UBC-MDS/Car_Evaluation_Analysis**
**Report link: https://ubc-mds.github.io/Car_Evaluation_Analysis/report/car_evaluation_analysis.html**
**Abstract/executive summary:**
This analysis aims to identify the key factors that determine whether a car is considered acceptable, good, or unacceptable according to standardized criteria. To achieve this, several common machine learning models were explored. The SVM RBF classifier was identified as the best-performing model, achieving a test accuracy of 0.952. The SVM RBF model demonstrated exceptional performance, with an accuarcy of 0.99 on unseen data. This makes the SVM RBF model a solid choice for this project.

The dataset used in this project is the Car Evaluation Database created by M. Bohanec and V. Rajkovic in the early 1990s. It was sourced from the UCI Machine Learning Repository and is publicly available for research and can be found in the [UCI Machine Learning Repository](https://archive.ics.uci.edu/dataset/19/car+evaluation).

**Editor:** @ttimbers 
**Reviewer:** Zhang, Yichi;  Singh, Paramveer; Chang, Yuyuan 

- [ ] I agree to abide by [MDS's Code of Conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by y1chi-z** [comment]
## Data analysis review checklist

### Reviewer: @y1chi-z

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robsutness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [ ] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [ ] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2

### Review Comments: 

Overall it is a great and well-organized project, you guys did a good job! Here are some comments:

- Some minor typos: for example, in the CONTRIBUTION.md file: Attribution: "TThese".
- Might want to include some limitations of this research in the methodology part.
- I notice that you have imbalanced dataset with a large portion of class `unacceptable`. Using accuracy to evaluate the model performance might not be enough. Trying other metrics like precison, recall or f1-score might be helpful.
- For the model optimization part, including the metric being used during the hyperparameter tuning would help us understand the process better, because the metric used directly influences the way the model is optimized.
- For the result part, including a confusion matrix plot can help us understand the result in a more straightforward way.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by Psingh238** [comment]
## Data analysis review checklist

### Reviewer: @Psingh238

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [x] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [x] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robustness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [ ] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [ ] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2

### Review Comments:

There are some missing details on how the various models were configured, such as the number of max neighbors for the *k*-NN and the max depth for the Decision Tree. These hyperparameters were also not subject to hyperparameter optimization. This was not explained in the report and thus the authors likely cannot dismiss these models outright without testing them thoroughly. I would be interested to see how these models compare to the SVM model when they have been optimized.

I really like the heatmap and the general use of images to convey information about the dataset. However, there was no comment by the authors on the imbalanced nature of the dataset and how it affected their analysis. I would like to know more about whether this could have led to the very high accuracy score for the test set as a 0.99 accuracy is higher than the cross-validation accuracy which could suggest a potential issue. Exploring this can lead to a great final conclusion which could lead to further insights for the future analysis.

Overall, I would say that this project is well organized and it is easy to see what each script does.

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by BChangs99** [comment]
## Data Analysis Review Checklist

### Reviewer: @bchangs99 

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation

- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelines:** Does the code adhere to well known language style guides?
- [ ] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robustness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [x] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [x] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 
2
### Review Comments: 
Firstly, this was a fantastically done report! The report flowed very well and I found the majority of the report very digestible to understand. I will note that I found a few grammatical errors here and there (for example the title of the report "Predicing..."  or in the Introduction "...as part of efforts..." ), but it largely did not detract from the report! 

As well, I wanted to mention that I read through your scripts and I found them all quite easy to follow along with all the comments. In terms of the testing checkbox above, I wasn't able to tick this off cause I wasn't sure if this criterion is specifically to unit tests for Milestone 4 or for the Data Validation testing. However, in terms of the modularity checkbox above, I think your validation steps in the split_n_preprocess.py could be pulled out as a separate script. Aside from this small critique I think everything else in your scripts is spectacularly well done!

That being said, I do have to agree with the feedback by @Psingh238 . Although I do see you mention the class imbalance from the large portion of `unacceptable` targets, I think this report could've benefited for example by expanding on it context-wise for why it may not have been important, or by examining/optimizing for f1-score. 

As well, I'll have to echo again his comment on hyperparameter optimization on the other models. I completely understand the computational and time cost of having to optimize for multiple different models, so I can empathize with why you decided to select just one model for optimization. However, I do agree that it wouldn't be right to label SVM as the best model without optimization on all models. That being said, I think just a disclaimer on why the other models weren't trained (e.g. computational/time cost) and a note in the discussions for how the project can be improved upon by optimizing these other models is enough!

In terms of other additional feedback, I noticed you've already finished your Makefile but haven't included instructions on how to run it in your [README.md](https://github.com/UBC-MDS/Car_Evaluation_Analysis/blob/main/README.md). Understandably this Milestone isn't finished, but I wanted to mention it just in case!

On that note, I also noticed your docker container doesn't include the Make dependency to be able to run the Makefile from within it. Through my own experience, I found out myself that the Make commands are able to run locally and not in the container because they come with xCode as a global dependency. I think adding a command in your Dockerfile to install will make the reproducibility of your report better!

One last constructive criticism, I completely understand that GitHub Pages wasn't a requirement for the Milestone, but I think your (and all) projects can be improved by setting it up to serve your rendered HTML to be more accessible. I'm mentioning this specifically as well as I can see you had it as a To-Do in your Issues [here](https://github.com/UBC-MDS/Car_Evaluation_Analysis/issues/17), but you closed it without resolving. 

You can setup GH pages by copying your rendered HTML report, renaming it to index.html and moving it into your root folder. Then by going into the settings in your GH, you can click on Pages on the left-hand side and turning on GH pages. If you need any help with this please feel free to reach out to me! This is just an added bonus for having your report be more accessible, I don't think you necessarily need to have this.

In general, I found this project and report very well done. Good work!

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.

**Comment by MarekB19** [comment]
## Data analysis review checklist

### Reviewer: @MarekB19 

### Conflict of interest

- [x] As the reviewer I confirm that I have no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [MDS code of conduct](https://ubc-mds.github.io/resources_pages/code_of_conduct/).

### General checks

- [x] **Repository:** Is the source code for this data analysis available? Is the repository well organized and easy to navigate?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?

### Documentation
- [x] **Installation instructions:** Is there a clearly stated list of dependencies? 
- [x] **Example usage:** Do the authors include examples of how to use the software to reproduce the data analysis?
- [x] **Functionality documentation:** Is the core functionality of the data analysis software documented to a satisfactory level?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Code quality
- [x] **Readability:** Are scripts, functions, objects, etc., well named? Is it relatively easy to understand the code?
- [x] **Style guidelides:** Does the code adhere to well known language style guides?
- [ ] **Modularity:** Is the code suitably abstracted into scripts and functions?
- [ ] **Tests:** Are there automated tests or manual steps described so that the function of the software can be verified? Are they of sufficient quality to ensure software robustness?

### Reproducibility
 
- [x] **Data:** Is the raw data archived somewhere? Is it accessible?
- [x] **Computational methods:** Is all the source code required for the data analysis available?
- [x] **Conditions:** Is there a record of the necessary conditions (software dependencies) needed to reproduce the analysis? Does there exist an easy way to obtain the computational environment needed to reproduce the analysis?
- [x] **Automation:** Can someone other than the authors easily reproduce the entire data analysis?

### Analysis report

- [x] **Authors:** Does the report include a list of authors with their affiliations?
- [x] **What is the question:** Do the authors clearly state the research question being asked?
- [x] **Importance:** Do the authors clearly state the importance for this research question?
- [ ] **Background**: Do the authors provide sufficient background information so that readers can understand the report?
- [x] **Methods:** Do the authors clearly describe and justify the methodology used in the data analysis? Do the authors communicate any assumptions or limitations of their methodologies?
- [x] **Results:** Do the authors clearly communicate their findings through writing, tables and figures?
- [ ] **Conclusions:** Are the conclusions presented by the authors correct? 
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?
- [x] **Writing quality:** Is the writing of good quality, concise, engaging? 

### Estimated hours spent reviewing: 2

### Review Comments: 

Fantastic work guys! Really great report, the points all flowed logically from one topic to another. The visuals were also great and well inserted in the report. The project folder is really well laid out, and I was able to follow along your scripts quite easily. The README file was great, I was able to ramp up and run your analysis easily. Seeing as I'm the last commenter here, most of the main points will have already been commented on (and in great details in the case of the comment above mine!). I have a few small nitpicks listed below, that you can choose to act on or not, But i figured i would point out what i see!

- [ ] I know this has likely been said before, but the 99% accuracy is something that immediately jumps out to me as extraordinarily high. In this case, data seems to have class imbalance. when paired with accuracy as scoring metrics will result in high even if the model happens to miss a relative high number of one of the smaller categories. This seems to be a multi-class classification problem, so i realize that techniques we learnt such as f1 score and confidence matrices may be harder to apply, but it might be worth considering in your conclusion if there was anything to do to address the imbalance, or at least discuss why accuracy was determined to still be the scoring metric and why its score might be so high.
- [ ] This is a tiny nitpick, but in the hyperparameter optimization script, i noticed you used random search with 100 iterations, but rather than picking from parameter distributions you listed discrete values. This is essentially the same as performing a grid search, since there is a pretty short list of possible parameter combinations. I really appreciated the hyperparameter grid the team created from these results however!
- [ ] In the report, the team did a great job introducing the problem, as well as enumerating the features that were used and their possible values. However, it wasn't immediately obvious to me what is the meaning of the target variable for this project (ie. what does "car acceptability" mean, is it based on client sentiment or some other metric, and what are the possible values you are trying to evaluate). it might be nice to just add a quick sentence or two regarding that.
- [ ] Another nitpick, but i found that some of the scripts had fairly generic names (ie evaluate_car_predictor vs evaluate_models). This made it a bit trickier for me to determine at first what each of these scripts was accomplishing. as well, while a full docstring is definitely not required, i found the current docstring for these two scripts was a bit generic, so i had to actually dig into the code to tell what each script accomplishes. 

Overall, aside from these points and  a few typos i noticed that were mentioned in the comments above, I found this project to be very professional and well executed. Congratulations on the work and best of luck in this final milestone! 

#### Attribution

This was derived from the [JOSE review checklist](https://openjournals.readthedocs.io/en/jose/review_checklist.html) and the ROpenSci review checklist.
